### PR TITLE
Core: Improvements around Token Refresh time expiration

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 1]
```diff
@@ -461,8 +463,9 @@ private OAuthTokenResponse refreshCurrentToken(RESTClient client) {
       }
     }
 
-    static boolean isExpired(Long expiresAtMillis, long now) {
-      return expiresAtMillis != null && expiresAtMillis < now;
+    static boolean isExpired(Instant expiresAt) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:466]: @rdblue you mentioned in https://github.com/apache/iceberg/pull/6489#discussion_r1072451113 that this type of method can be misleading, because the token could still be expired even though `expiresAt` is `null`.
To me it makes sense to have this method, given that we're using it at 2 different places. 
Technically we're just checking whether a given `Instant` is expired or not so IMO it's clear what the purpose of this method is. 
I can't come up with a better name, but I'm happy to rename it to something more obvious in case you have suggestions. 
Would `isNonNullAndExpired()` be more descriptive about the purpose of the method?

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 2]
```diff
@@ -126,8 +127,7 @@ public void initialize(String name, Map<String, String> unresolved) {
     // catalog service
     Map<String, String> props = EnvironmentUtil.resolveAll(unresolved);
 
-    long startTimeMillis =
-        System.currentTimeMillis(); // keep track of the init start time for token refresh
+    Instant startTime = Instant.now(); // keep track of the init start time for token refresh
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:130]: Throughout Iceberg, we use `System.currentTimeMillis()`. Is there a reason to deviate from that?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:466]: I don't think we need a new method just because we do a null check twice. The null check and comparison is clear and readable.

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:466]: it is less about the null check and more about the expiration check that we repeat at 2 places

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:130]: calculations with an `Instant` are generally easier. Additionally we don't have to add `xyzMillis` to every variable to indicate that those are milliseconds

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 3]
```diff
@@ -452,7 +452,7 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
     }
 
     private OAuthTokenResponse refreshCurrentToken(RESTClient client) {
-      if (isExpired(expiresAtMillis, System.currentTimeMillis())) {
+      if (expiresAtMillis != null && expiresAtMillis <= System.currentTimeMillis()) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:455]: @rdblue note that we're using `<=` rather than `<` here

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java]: I've removed the method and reverted the code to how it was before

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 4]
```diff
@@ -550,31 +512,29 @@ public static AuthSession fromAccessToken(
...
           new AuthSession(
               parent.headers(), token, OAuth2Properties.ACCESS_TOKEN_TYPE, parent.credential());
 
-      long startTimeMillis = System.currentTimeMillis();
-      Long expiresAtMillis = session.expiresAtMillis();
+      Long expiresAtMillis = null;
 
-      if (isExpired(expiresAtMillis, startTimeMillis)) {
+      if (session.expiresAtMillis() != null
+          && session.expiresAtMillis() <= System.currentTimeMillis()) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:524]: Why did you change the use of `expiresAtMillis`? Now there are 4 calls to `session.expiresAtMillis()` rather than using a variable.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 5]
```diff
@@ -550,31 +512,29 @@ public static AuthSession fromAccessToken(
...
+        long startTimeMillis = System.currentTimeMillis();
         Pair<Integer, TimeUnit> expiration = session.refresh(client);
         if (expiration != null) {
-          scheduleTokenRefresh(
-              client, executor, session, startTimeMillis, expiration.first(), expiration.second());
+          expiresAtMillis = startTimeMillis + expiration.second().toMillis(expiration.first());
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:528]: Now that token refresh is based around `expiresAt`, should we prefer the specific timestamp from the token if it exists? I think it would make sense to check `session.expiresAtMillis()` here and use it in place of the duration, since the duration requires careful handling of start time.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 6]
```diff
@@ -550,31 +512,29 @@ public static AuthSession fromAccessToken(
...
               parent.headers(), token, OAuth2Properties.ACCESS_TOKEN_TYPE, parent.credential());
 
-      long startTimeMillis = System.currentTimeMillis();
-      Long expiresAtMillis = session.expiresAtMillis();
+      Long expiresAtMillis = null;
 
-      if (isExpired(expiresAtMillis, startTimeMillis)) {
+      if (session.expiresAtMillis() != null
+          && session.expiresAtMillis() <= System.currentTimeMillis()) {
+        long startTimeMillis = System.currentTimeMillis();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:525]: Moving this variable causes two calls to `currentTimeMillis()`. Can you move it back?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 7]
```diff
@@ -611,17 +571,17 @@ private static AuthSession fromTokenResponse(
...
-            session,
-            startTimeMillis,
-            response.expiresInSeconds(),
-            TimeUnit.SECONDS);
+
+        expiresAtMillis = startTimeMillis + TimeUnit.SECONDS.toMillis(response.expiresInSeconds());
       } else if (session.expiresAtMillis() != null) {
-        scheduleTokenRefresh(
-            client, executor, session, session.expiresAtMillis(), TimeUnit.MILLISECONDS);
+        expiresAtMillis = session.expiresAtMillis();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:580]: I think we should start to prefer the token's expiration if it exists. Can you swap the priority here?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 8]
```diff
@@ -611,17 +571,17 @@ private static AuthSession fromTokenResponse(
...
+      Long expiresAtMillis = null;
       if (response.expiresInSeconds() != null) {
-        scheduleTokenRefresh(
-            client,
-            executor,
-            session,
-            startTimeMillis,
-            response.expiresInSeconds(),
-            TimeUnit.SECONDS);
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:577]: Nit: unnecessary newline.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 5 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:528]: that makes sense

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 7 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:580]: done

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 5 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java]: I've updated the code and noticed that we probably still have to check both: the expiration time from the refreshed token (if it's set and if token refresh was successful), and then fall back to the expiration that we get from the response.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 9]
```diff
@@ -585,22 +547,24 @@ public static AuthSession fromAccessToken(
       long startTimeMillis = System.currentTimeMillis();
       Long expiresAtMillis = session.expiresAtMillis();
 
-      if (isExpired(expiresAtMillis, startTimeMillis)) {
+      if (null != expiresAtMillis && expiresAtMillis <= startTimeMillis) {
         Pair<Integer, TimeUnit> expiration = session.refresh(client);
+        // if expiration is non-null, then token refresh was successful
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:552]: I think we need an else case that avoids scheduling token refresh. If the expiration is null then the refresh was not successful and should not be reattempted. But in that case, `expiresAtMillis` has the original expired value and will cause the last check to schedule a token refresh.

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:552]: good catch, that makes sense. I've pushed a fix where we reset the original `expiresAtMillis` when the token refresh failed

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @nastra!
